Fix issue where full-SHA alias is not supported for links with user defined networks

yongtang opened the pull request:
**- What I did**


This fix tries to address the issue raised in #35573 where full-SHA alias is not supported for links with user defined networks.


**- How I did it**

The reason was that CreateOptionAlias used name directly but CreateOptionMyAlias used short ID of the container to alias.

This fix use short ID for CreateOptionAlias so that the issue could be addressed.

**- How to verify it**

A test case has been added to cover the changes.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

![unnamed](https://user-images.githubusercontent.com/6932348/34229333-bb5f22e8-e588-11e7-8387-7c0d55cbbda6.jpg)

This fix fixes #35573.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

thaJeztah commented:
Looks like there's failures in `docker-py` tests;

```
02:37:42 =================================== FAILURES ===================================
02:37:42 _____________________ TestNetworks.test_connect_with_links _____________________
02:37:42 /docker-py/tests/integration/api_network_test.py:338: in test_connect_with_links
02:37:42     links=[('docker-py-test-upstream', 'bar')])
02:37:42 /docker-py/docker/utils/decorators.py:19: in wrapped
02:37:42     return f(self, resource_id, *args, **kwargs)
02:37:42 /docker-py/docker/utils/decorators.py:34: in wrapper
02:37:42     return f(self, *args, **kwargs)
02:37:42 /docker-py/docker/api/network.py:253: in connect_container_to_network
02:37:42     self._raise_for_status(res)
02:37:42 /docker-py/docker/api/client.py:224: in _raise_for_status
02:37:42     raise create_api_error_from_http_exception(e)
02:37:42 /docker-py/docker/errors.py:31: in create_api_error_from_http_exception
02:37:42     raise cls(e, response=response, explanation=explanation)
02:37:42 E   NotFound: 404 Client Error: Not Found ("No such container: docker-py-test-upstream")
02:37:42 _____________________ TestNetworks.test_create_with_links ______________________
02:37:42 /docker-py/tests/integration/api_network_test.py:303: in test_create_with_links
02:37:42     links=[('docker-py-test-upstream', 'bar')],
02:37:42 /docker-py/tests/integration/base.py:117: in create_and_start
02:37:42     self.client.start(container)
02:37:42 /docker-py/docker/utils/decorators.py:19: in wrapped
02:37:42     return f(self, resource_id, *args, **kwargs)
02:37:42 /docker-py/docker/api/container.py:1085: in start
02:37:42     self._raise_for_status(res)
02:37:42 /docker-py/docker/api/client.py:224: in _raise_for_status
02:37:42     raise create_api_error_from_http_exception(e)
02:37:42 /docker-py/docker/errors.py:31: in create_api_error_from_http_exception
02:37:42     raise cls(e, response=response, explanation=explanation)
02:37:42 E   NotFound: 404 Client Error: Not Found ("No such container: docker-py-test-upstream")
```

thaJeztah commented on integration/network/link_test.go:
```diff
@@ -0,0 +1,100 @@
+
+	body, err := client.ContainerLogs(ctx, "bar", types.ContainerLogsOptions{
+		ShowStdout: true,
+		//ShowStderr: true,
```
can you remove this one?

yongtang replied:
Thanks @thaJeztah. The PR has been updated.

yongtang commented:
@thaJeztah The PR has been updated. Please take a look.

MHBauer commented on daemon/container_operations.go:
```diff
@@ -768,9 +768,25 @@ func (daemon *Daemon) connectToNetwork(container *container.Container, idOrName
+			// will be used if there is no container exists
+			if child, err := daemon.GetContainer(name); err == nil {
+				name = stringid.TruncateID(child.ID)
+			}
```
The comment makes it look like these three lines are preserving the old behavior, but this is the implementation of the fix that turns a `name`, which coming out of `ParseLink` is really a full long ID.

I don't even remotely know how libnetwork works to suggest how the comment should be adjusted. Maybe remove the content before the comma, and leave only something like "use truncated container id if linked container was a name"

MHBauer reviewed (commented):
I am having a bit of trouble with the comments, but I think the code looks okay.

I would like to see additional test cases beyond full ID to the other cases of name and shrort ID, unless those are tested elsewhere.

Remind me to come back around and look at this again.

yongtang commented:
@MHBauer Thanks for the review. The PR has been updated with additional tests as well as the updated comments. Please take a look.

MHBauer commented on integration/network/link_test.go:
```diff
@@ -0,0 +1,102 @@
+	err = client.ContainerStart(ctx, foo.ID, types.ContainerStartOptions{})
+	require.NoError(t, err)
+
+	aliases := []string{foo.ID, foo.ID[:5], "foo"}
```
nit,add comment
// these are the test cases of full id,partial id, and container name.

yongtang replied:
@MHBauer Thanks. The comment has been updated.

MHBauer commented:
looks good to me, do we need @dnephin or @vdemeester to take a look? I've never interacted with code owners before.

yongtang commented:
Ping @dnephin  @vdemeester, please take a look.

dnephin commented on daemon/container_operations.go:
```diff
@@ -775,9 +775,25 @@ func (daemon *Daemon) connectToNetwork(container *container.Container, idOrName
-	joinOptions, err := container.BuildJoinOptions(n)
-	if err != nil {
-		return err
+	// Builds endpoint Join options from a given network.
```
This could still be a separate function in this package.

dnephin commented on integration/network/link_test.go:
```diff
@@ -0,0 +1,103 @@
+package network
```
It should be possible to test this behaviour without an integration test

dnephin commented:
code owners are just being used as a way to get notified when a PR changes some file that matches a pattern. It doesn't necessary mean the PR needs approval from one.

yongtang commented:
@dnephin It might be difficult to come up with a unit test as:
https://github.com/yongtang/docker/blob/9fc57f1c48614668668fa3290cdcf10ee9fb1830/daemon/container_operations.go#L787-L791
```
			// To preseve old behavior, use truncated container id
			// if linked container was a name.
			if child, err := daemon.GetContainer(name); err == nil {
				name = stringid.TruncateID(child.ID)
			}
```
needs to interact with daemon (`daemon.GetContainer(name)`) to resolve the linked container. Not sure what is the best way to handle this case.

dnephin commented:
If you extract the code back into a function, the function could accept an interface with `GetContainer()`. In the test you can pass a fake that implements the interface to return the value you want for that case.

yongtang commented:
Thanks @dnephin After taking another look I think the function parameter of `GetContainer` will help.

But there are still other issues. At the moment the
```
BuildJoinOptions(n named) ([]libnetwork.EndpointOption, error) {
```
returns `EndpointOption` which is an option setter function type with unexpected parameter types:
```
type EndpointOption func(ep *endpoint)
```
It is really quite difficult to test against `EndpointOption` it looks.

dnephin commented:
Ah yes, that does make it very difficult to test. I'm out of ideas

yongtang commented:
@dnephin Maybe we add the integration test for now, to have the fix for #35573? It may not be ideal though that might be an option at the moment.

vdemeester commented:
ping @yongtang what is the status here ? also need a rebase :sweat_smile:

yongtang committed:
Fix issue where full-SHA alias is not supported for links with user defined networks

This fix tries to address the issue raised in 35573 where
full-SHA alias is not supported for links with user defined networks.

The reason was that CreateOptionAlias used name directly but CreateOptionMyAlias
used short ID of the container to alias.

This fix use short ID for CreateOptionAlias so that the issue could
be addressed.

This fix fixes 35573.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for full-SHA alias supported of links with user defined networks

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @vdemeester. The PR has been rebased. The integration test is included. Please take a look.

vdemeester reviewed (approved):
SGTM :camel: 
cc @thaJeztah @fcrisciani

thaJeztah commented:
Looks like a function needs to be removed (or the new code put in that function);

```
07:54:44 daemon/network.go:900:1:warning: buildJoinOptions is unused (deadcode)
07:54:44 Build step 'Execute shell' marked build as failure
```

yongtang commented:
@vdemeester @thaJeztah The tests passed except powerpc which I think is unrelated (#37277).
```
02:40:23 FAIL: docker_cli_daemon_test.go:1800: DockerDaemonSuite.TestDaemonNoSpaceLeftOnDeviceError
02:40:23 
02:40:23 [dcfb192c0bd25] waiting for daemon to start
02:40:23 [dcfb192c0bd25] exiting daemon
02:40:23 docker_cli_daemon_test.go:1817:
02:40:23     s.d.Start(c, "--data-root", filepath.Join(testDir, "test-mount"))
02:40:23 /go/src/github.com/docker/docker/internal/test/daemon/daemon.go:203:
02:40:23     t.Fatalf("Error starting daemon with arguments: %v", args)
02:40:23 ... Error: Error starting daemon with arguments: [--data-root /tmp/no-space-left-on-device-test032730539/test-mount]
```

yongtang commented:
@thaJeztah @vdemeester All tests passed now. Please take a look.

vdemeester reviewed (approved):
Still LGTM :stuck_out_tongue:

olljanat commented:
@thaJeztah PTAL

piyush-insider commented:
Hi, @yongtang @vdemeester is this being actively worked on? Is it safe to close this PR now?

yongtang commented:
@thaJeztah Is an integration test OK for this PR? I can rebase and update the PR if this is the case.

akerouanton commented:
As we're in the process of fully removing legacy links (see this RFC: https://github.com/moby/moby/issues/47464), let me close this PR.
